### PR TITLE
Possibility to change S3 disk settings in runtime

### PR DESCRIPTION
**Pull request by Jokser**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Possibility to change S3 disk settings in runtime via new `SYSTEM RESTART DISK` SQL command.

Detailed description / Documentation draft:
Introduced DiskRestartProxy that re-applies changed S3 disk settings in runtime.
Introduced `SYSTEM RESTART DISK` SQL command that invokes DiskRestartProxy->restart().
Reduced memory usage for WriteBufferFromS3 decorators (each decorator allocated own buffer in memory that was not actually used).
ReadBufferFromS3/WriteBufferFromS3 decorators code refactoring.

### CODE REVIEW

File: src/Disks/DiskRestartProxy.cpp [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+{
+public:
+    RestartAwareReadBuffer(const DiskRestartProxy & disk, std::unique_ptr<ReadBufferFromFileBase> impl_)
+        : ReadBufferFromFileDecorator(std::move(impl_)), lock(disk.mutex) { }
+
+private:
+    ReadLock lock;
+};
+
+/// Holds restart write lock till finalize.
```
Comment (Jokser) [inline src/Disks/DiskRestartProxy.cpp:27]: Typo. There should be 'read lock'

File: src/IO/ReadBufferFromFileDecorator.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+{
+    return impl->getFileName();
+}
+
+
+off_t ReadBufferFromFileDecorator::getPosition()
+{
+    swap(*impl);
+    auto position = impl->getPosition();
+    swap(*impl);
```
Comment (vdimir) [inline src/IO/ReadBufferFromFileDecorator.cpp:24]: I see that this swap was in deleted code, but didn't get why we swap buffers

Comment (Jokser) [reply src/IO/ReadBufferFromFileDecorator.cpp:24]: Decorator buffers don't have own buffer in memory, so we need to 'steal' actual working buffer from `impl`, perform an operation and return buffer back.

File: src/Disks/S3/DiskS3.cpp [hunk 3]
```diff
@@ -1518,4 +1499,40 @@ void DiskS3::onFreeze(const String & path)
...
+{
+    auto new_settings = disk_settings_getter(context->getConfigRef(), "storage_configuration.disks." + name, context);
+
+    client = new_settings.client;
+    s3_max_single_read_retries = new_settings.s3_max_single_read_retries;
+    s3_min_upload_part_size = new_settings.s3_min_upload_part_size;
+    s3_max_single_part_upload_size = new_settings.s3_max_single_part_upload_size;
+    min_bytes_for_seek = new_settings.min_bytes_for_seek;
+    send_metadata = new_settings.send_metadata;
+    list_object_keys_size = new_settings.list_object_keys_size;
```
[the remark this answers is not in the document]
Comment (Jokser) [reply src/Disks/S3/DiskS3.cpp:1512]: Okay, makes sense.

File: src/Disks/IDisk.h [hunk 4]
```diff
@@ -211,24 +212,33 @@ class IDisk : public Space
...
     virtual void onFreeze(const String &) { }
 
     /// Returns guard, that insures synchronization of directory metadata with storage device.
     virtual SyncGuardPtr getDirectorySyncGuard(const String & path) const;
 
+protected:
+    friend class DiskDecorator;
+
+    /// Applies new settings for disk in runtime.
+    virtual void applyNewSettings(ContextConstPtr) { }
```
[the remark this answers is not in the document]
Comment (Jokser) [reply src/Disks/IDisk.h:237]: Okay, makes sense

**Review by vdimir — APPROVED** [review]

**Comment by sevirov** [comment]
Internal documentation ticket: DOCSUP-18488

**Comment by HeenaBansal2009** [comment]
@sevirov , I was wondering for the documentation for this feature. 
I will appreciate if you can share the link relate dot this feature and was delivered asn part of Internal documentation ticket: DOCSUP-18488.

Thanks!

**Comment by vdimir** [comment]
@HeenaBansal2009 I'm afraid it wasn't finished and is abandoned. I suppose best way to understand how feature work is to read changes in integration test for this PR.

**Comment by HeenaBansal2009** [comment]
Thanks @vdimir for response. However , I am just curious Since this feature was implemented for backup/restore metadata files from S3 ?  If it is not production use and abandoned , How current backup/restore feature of CH  are going to be affected . 

I was searching for any backup restore documentation if available. 

Thanks in advance.

**Comment by vdimir** [comment]
@HeenaBansal2009 New backup/restore is not related to this ([here](https://github.com/ClickHouse/ClickHouse/pulls?q=is%3Apr+backup+is%3Aclosed+author%3Avitlibar+sort%3Acreated-asc) is the PRs for new backup/restore). Not sure about docs for the new feature. @vitlibar maybe you can answer?

**Comment by HeenaBansal2009** [comment]
@vdimir Its good to see that backup/restore support is planned for 2022. I did find  docs in clickhouse documentation for backup/restore commands.

However , I wanted to know  is these commands does online backup? OR  backup command does freeze the table and take snapshot

**Comment by Slach** [comment]
@HeenaBansal2009 https://clickhouse.com/docs/en/operations/backup/ some documentation here
backup make copy snapshot to destination

**Pull request merged**